README
CardapioBandex
Copyright (c) 2009 Raphael Kubo da Costa <kubito@gmail.com>
===========================================================

This is a Python library to parse State University of Campinas' (www.unicamp.br) university restaurant's menu (http://www.prefeitura.unicamp.br/servicos.php?servID=119).

It can be used to then send the menu via SMS or e-mail to students or do whatever kind of crazy manipulation you want to.

This library depends on BeautifulSoup (version 3+, I guess), which can be obtained from http://www.crummy.com/software/BeautifulSoup.


debian/changelog
python-cardapiobandex (0.1.0-1) unstable; urgency=low

  * Initial release.

 -- Raphael Kubo da Costa <kubito@gmail.com>  Sat, 28 Mar 2009 14:36:47 -0300


example.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

#
# Copyright (c) 2009, Raphael Kubo da Costa <kubito@gmail.com>
#
# Permission to use, copy, modify, and/or distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#

"""
This is a sample program to show what can be done with the library.
"""

from CardapioBandex import Parser

if __name__ == "__main__":
  app = Parser.Parser()
  menu = app.parseMenu()

  # The full-text version of the menu
  menu.printMenu()

  # A compact version, useful for using in SMS
  menu.printSMSMenu()


cardapio-bandex.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

#
# Copyright (c) 2009, Raphael Kubo da Costa <kubito@gmail.com>
#
# Permission to use, copy, modify, and/or distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#

from CardapioBandex import Parser

if __name__ == "__main__":
  app = Parser.Parser()

  menu = app.parseMenu()
  menu.printMenu()
  menu.printSMSMenu()


bandejao-unicamp.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from BeautifulSoup import BeautifulSoup, BeautifulStoneSoup
from xml.sax.saxutils import unescape
import urllib2
import re

class Menu(object):
  def __init__(self, values):
    self.fromDict(values)

  def fromDict(self, values):
    if not isinstance(values, dict):
      raise TypeError, "A dict object must be passed to Menu"

    if set(values.keys()) != set(['data', 'principal', 'sobremesa', 'salada',
                                  'suco']):
      raise ValueError, "Malformed dictionary passed to Menu"

    self.values = values

  def printMenu(self):
    for key in ('data', 'principal', 'salada', 'sobremesa', 'suco'):
      print '%s: %s' % (key, self.values[key])

  def printSMSMenu(self):
    ## Date
    data_re = re.match(r'(?:(.+) - )?([\d]{1,2}/[\d]{1,2})(?:\d{2,4})*',
                       self.values['data'])

    if not data_re:
      raise ValueError, "Invalid date format"

    if len(data_re.groups()) == 2:
      short_data = data_re.group(1)[:3] + ' ' + data_re.group(2)
    else:
      short_data = data_re.group(1)

    ## Principal
    short_principal = self.values['principal'].lower(). \
                                               replace(u'feijão', 'f'). \
                                               replace('arroz', 'a')

    ## Salada
    short_salada = self.values['salada'].lower()

    ## Sobremesa
    short_sobremesa = self.values['sobremesa'].lower()

    ## Suco
    short_suco = self.values['suco'].lower()

    ## Final message
    short_message = "%s - %s/Sl: %s/Sb: %s/Su: %s" % (short_data,
                                                      short_principal,
                                                      short_salada,
                                                      short_sobremesa,
                                                      short_suco)
    short_message = short_message.replace(', ', ',').replace('.', '')

    print len(short_message), short_message

class MenuParser(object):
  SOURCE_FIXES = [
    # <td background-color: "#ededed;"> becomes <td>
    (re.compile(r'\s*[-\w]+:\s*"[a-zA-Z0-9#;]+"'), lambda s: ''),
  ]
  SOURCE_URL = "http://www.prefeitura.unicamp.br/servicos.php?servID=119"

  def __init__(self):
    self.sourceText = self.getSource()

  def getSource(self):
    return urllib2.urlopen(MenuParser.SOURCE_URL).read()

  def parseMenu(self):
    soup = BeautifulSoup(self.sourceText,
                         markupMassage = MenuParser.SOURCE_FIXES,
                         convertEntities = BeautifulStoneSoup.HTML_ENTITIES)

    menuBlock  = soup.findAll('p', style="text-align: left")
    menuKeys   = ('data', 'principal', 'sobremesa', 'salada', 'suco')
    menuValues = []

    for line in menuBlock:
      strLine = ''.join(line.findAll(text=True)).replace(unichr(160), ' ')
      match   = re.match(r'[\w\s]+:\s*(.+)', strLine)

      if match:
        menuValues.append(match.group(1))

    return Menu(dict(zip(menuKeys, menuValues)))

if __name__ == "__main__":
  app = MenuParser()

  menu = app.parseMenu()
  menu.printMenu()
  menu.printSMSMenu()


CardapioBandex/Menu.py
# -*- coding: utf-8 -*-

#
# Copyright (c) 2009, Raphael Kubo da Costa <kubito@gmail.com>
#
# Permission to use, copy, modify, and/or distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#

import re

class Menu(object):
  def __init__(self, values):
    self.fromDict(values)

  data      = property(lambda self: self.values['data'])
  principal = property(lambda self: self.values['principal'])
  salada    = property(lambda self: self.values['salada'])
  sobremesa = property(lambda self: self.values['sobremesa'])
  suco      = property(lambda self: self.values['suco'])

  def fromDict(self, values):
    """
    Validates and loads a dict with the menu values
    """
    if not isinstance(values, dict):
      raise TypeError, "A dict object must be passed to Menu"

    if set(values.keys()) != set(['data', 'principal', 'sobremesa', 'salada',
                                  'suco']):
      raise ValueError, "Malformed dictionary passed to Menu"

    self.values = values

  def printMenu(self):
    """
    Prints the menu in the usual, multi-line way.
    """
    for key in ('data', 'principal', 'salada', 'sobremesa', 'suco'):
      print '%s: %s' % (key, self.values[key])

  def printSMSMenu(self):
    """
    Prints the menu in a format suitable for SMS messages.
    """
    ## Date
    data_re = re.match(r'(?:(.+)\s*-\s*)?([\d]{1,2}/[\d]{1,2})(?:\d{2,4})*',
                       self.values['data'])

    if not data_re:
      raise ValueError, "Invalid date format"

    if data_re.group(1) is not None:
      short_data = data_re.group(1)[:3] + ' ' + data_re.group(2)
    else:
      short_data = data_re.group(2)

    ## Principal
    short_principal = self.values['principal'].lower(). \
                                               replace(u'feijão', 'f'). \
                                               replace('arroz', 'a')

    ## Salada
    short_salada = self.values['salada'].lower()

    ## Sobremesa
    short_sobremesa = self.values['sobremesa'].lower()

    ## Suco
    short_suco = self.values['suco'].lower()

    ## Final message
    short_message = "%s - %s/Sl: %s/Sb: %s/Su: %s" % (short_data,
                                                      short_principal,
                                                      short_salada,
                                                      short_sobremesa,
                                                      short_suco)
    short_message = short_message.replace(', ', ',').replace('.', '')

    print len(short_message), short_message


CardapioBandex/Parser.py
# -*- coding: utf-8 -*-

#
# Copyright (c) 2009, Raphael Kubo da Costa <kubito@gmail.com>
#
# Permission to use, copy, modify, and/or distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#

try:
  from BeautifulSoup import BeautifulSoup, BeautifulStoneSoup
except ImportError:
  import sys
  sys.stderr.write("This program needs BeautifulSoup to run.\n")
  sys.exit(1)

import urllib2
import re
import Menu

class Parser(object):
  SOURCE_FIXES = [
    # <td background-color: "#ededed;"> becomes <td>
    (re.compile(r'\s*[-\w]+:\s*"[a-zA-Z0-9#;]+"'), lambda s: ''),
  ]
  SOURCE_URL = "http://www.prefeitura.unicamp.br/servicos.php?servID=119"

  def __init__(self, source = None):
    if source is None:
      self.sourceText = self.getSource().read()
    else:
      self.sourceText = source.read()

  def __getTagText(self, tag):
    """
    Receives a Tag object and returns its stripped text.
    """
    return ''.join(tag.findAll(text=True)).replace(unichr(160), ' ')

  def getSource(self):
    """
    Returns the HTML file that should be parsed.
    """
    return urllib2.urlopen(Parser.SOURCE_URL)

  def parseMenu(self):
    """
    Parses an HTML source.
    """
    soup = BeautifulSoup(self.sourceText,
                         markupMassage = Parser.SOURCE_FIXES,
                         convertEntities = BeautifulStoneSoup.HTML_ENTITIES)

    keyTranslations = {'prato principal': 'principal',
                       'salada': 'salada',
                       'sobremesa': 'sobremesa',
                       'suco': 'suco'}
    menu = {}

    # The code here is not very smart or efficient, but we are trying
    # to be as much error-proof as possible.
    # Using the main meal and reading the other entries using it as a
    # reference point seems to be the best take.
    # We begin by finding the main meal, parsing the date and then the
    # meal items.
    mainMeal = soup.find(text = re.compile(r"prato\s+principal", re.IGNORECASE)).parent.parent
    mealDate = mainMeal.findPreviousSibling('p')
    menu['data'] = self.__getTagText(mealDate)

    # The four remaining entries (main meal, dessert, salad and juice) are the 4 next siblings
    for item in mealDate.findNextSiblings('p', limit=4):
      match = re.match(r'\s*([\w\s]+):\s*(.+)', self.__getTagText(item))

      if match:
        key = keyTranslations[match.group(1).lower()]
        menu[key] = match.group(2)

    # In the end, we shall have a dictionary with the following keys:
    # 'data', 'principal', 'salada', 'sobremesa', 'suco'
    if len(menu) != 5:
      raise ValueError, "The menu in the site has an invalid format"

    return Menu.Menu(menu)
